### ISSUE TITLE
Suggestion: Introduce HTTP Problem API

### ISSUE THREAD

**Issue by etiennedi** [semi-technologies/semi-website#485]
I wasn't sure whether to open this on semi-technologie/weaviate or on semi-technologies/semi-website as the impact would be on both. I've decided to go for this as the impact here would probably be bigger.

## Background
Sometimes we return error messages in weaviate that are really long and require a lot of explanation. For example [this error message](https://github.com/semi-technologies/weaviate/blob/beb55b57560f0895f956fe61b38c2d53bed598d7/usecases/schema/validation.go#L281-L286). Since it's just a string in a json object formatting is also hard. We can explicitly set `\n` line breaks, but the user would first have to translate them to literal line breaks to see the content formatted that way.

While looking for standards for interlinking API resources for semi-technologies/weaviate#1103 I came across [RFC7807](https://tools.ietf.org/html/rfc7807) which I think would be perfect for Weaviate.

## Suggestion

The idea is that we define so called HTTP problem. A problem is essentially more fine grained than a status code. So instead of returning `400 Bad Request`, we can return `400` and a link to an HTTP problem.

So for the error above we could return something like
```
HTTP/1.1 400 Bad Request
Content-Type: application/problem+json
Content-Language: en

{
  "status": 400,
  "type": "https://semi.technology/documentation/weaviate/problems/no-vectorization-guarantee"
  "title": "Not enough indexed fields to guarantee vectorization",
  "detail": "You have have disabled `vectorizeClassName`, but the class has no other indexed fields."
}
```

The import line is the `"type"` line. This would be an immutable link (it can redirect to another page, but the link must be immutable), to a page in the docs that describes exactly this problem (and only this problem). The problem pages can, but don't have to show up in navigation, but it is also fine if they are only discoverable through the API error message.

## Why I think this is a good idea

1. It's following a standard (RFC7807)
1. It gives us the freedom to have rich error messages, properly formatted with all kinds of content. Possibly even video in extreme cases
1. The actual json message the user sees will be reduced to a title and a short detail, as opposed to a long and awkward string.
1. It's cheap to implement and we could start with the most common ones to further reduce the effort.

cc @fefi42 @laura-ham @bobvanluijt What do you think?

**Comment by laura-ham** [comment]
Sounds like a really good idea, all for it. I don't have anything to add / propose regarding this at the moment. 

One question: Would you want to have this page `"https://semi.technology/documentation/weaviate/problems/"` hidden on the website or as part of the docs so people can already see them? Just a detail, but this question will come when we implement it.

**Comment by etiennedi** [comment]
> Would you want to have this page "https://semi.technology/documentation/weaviate/problems/" hidden on the website or as part of the docs so people can already see them? Just a detail, but this question will come when we implement it.

To be honest, I don't know yet. I see reasons both against and for it. Having them visible users who prefer a more theoretical approach (i.e. try to read the docs like a book before ever trying a query) will have the chance to learn more about Weaviate. Especially the example above ("no guaranteed vectorized fields") tells the user quite a bit about how vectorization works internally. In a way the List of Problems could be indexed like an FAQ ("How do I solve ....")

On the other hand just seeing a list of possible problems might also be confusing for readers who don't have the context of just sending a failing request <-- This could probably be mitigated by structuring the Problem page in a way where we first show an example request that would lead to the problem, then explain it in more detail, then present the solution.

After writing this, I'm more inclined to say "show it, don't hide it" on the website.

**Comment by bobvanluijt** [comment]
I agree with both of you, we can just add it to the docs in a visible manner.

I would like to add it to the current structure though, i.e., `https://semi.technology/documentation/weaviate/current/problems/`

Some questions/suggestions:
- the word `problems` is negative in tone. Can't we use another positive term (e.g., `solutions`) or a neutral one?
- Can we do it based on numbers? So that a 'problem' has a number. E.g., `https://semi.technology/documentation/weaviate/current/problems/0001.html` this makes it easy to test if all problems are part of the website. I could even imagine, that if @etiennedi adds a new 'problem'  number to Weaviate, it automatically* pushes an issue to the website that says "create issue 0010".

*- if it can even automatically assign @laura-ham would be even more awesome 😇😇😇

**Comment by fefi42** [comment]
Can we somehow track how many people are coming to the error page from a weaviate output directly? So we know what error is experienced the most and we can improve the docs for it or maybe even add additional checks to prevent it in the first place.
Maybe by adding some kind of unique id in the parameters that we only count once.

**Comment by etiennedi** [comment]
Some thoughts:

* I don't think it's ideal to put it at `current` because that changes over time. If someone runs an old version of weaviate it might no longer contain the problem the user is experiencing. This is not however an issue if we make sure that we never delete problems themselves. (It's fine to delete them from the FAQ/index page) 
* I prefer the word `problem` because it's explicitly defined in the standard. The term `solution` would also send the wrong signal in my opinion. Because then the definition of a problem stays in weaviate and we only link to a solution. That is not the idea behind RFC7807. There the idea is that a problem is globally defined as its own resource. When ever weaviate returns a problem, it basically creates an instance of that problem (possibly with a `details` field that makes that instance specific). There is also a 1..n relationship. A problem might have more than one solution. 
* I think we must keep file endings, such as `0001.html`, out of the problem defintion URI. If the website requires this structure there should be a redirect. This is because the problem is a REST resource. The fact that we return html is just one specific way to view this problem, but not part of the problem resource itself. (For example a problem could also be viewed via a Smartphone app where it wouldn't return html, but something suitable for the app). 
* In my opinions numbers have more downsides than upsides: The resource is no longer descriptive, it's impossible to tell what it means just from looking at it. To check if all problems are present we could just aggregate all problems in weaviate by id (Note: id is the human-readable string) and automatically check if it exists on the website. If it returns 404, we know it doesn't. Similarly we can still create issues (either manually, semi-automatically or entirely automatically) based on new problems that are added. I think semi-manually is the best option here, as this makes sure we don't forget about new problems, but still gives me the opportunity to describe the problem, so that Laura can turn my description into a nicely-formulated text.

**Comment by etiennedi** [comment]
> Can we somehow track how many people are coming to the error page from a weaviate output directly?

A pragmatic approach would be to assume that anyone that doesn't have the website overview page as a referrer set is coming to the page because they experienced the problem.

**Comment by bobvanluijt** [comment]
@etiennedi where could I find a list of all potential errors?

**Comment by etiennedi** [comment]
> @etiennedi where could I find a list of all potential errors?

Such a list does not exist yet, as errors aren't centralized at the moment.

**Comment by bobvanluijt** [comment]
Aha, would it be an idea to do this in a YAML way?

E.g.,

```yml
--- 
- 
  id: some-error
  message: "Bacon ipsum dolor amet tri-tip swine sausage brisket boudin strip steak. Kevin chuck meatloaf tri-tip kielbasa ground round strip steak bacon ball tip spare ribs t-bone prosciutto tail."
- 
  id: another-error
  message: "Tenderloin alcatra ball tip picanha strip steak chislic. Pork belly meatloaf shankle prosciutto pork pancetta."
```

This would result in:
- `//semi.technology/documentation/weaviate/problems/some-error.html`
- `//semi.technology/documentation/weaviate/problems/another-error.html`

---

@fefi42 just noticed that I didn't answer `Can we somehow track how many people are coming to the error page from a weaviate output directly?` we have Google Analytics to track this on the website.

**Comment by etiennedi** [comment]
I wouldn't want to track it in a `yaml` file as it's quite complex to read in a yaml file at compile time in Go. However, I can generate a `yaml` file from the go code for the website if you like :)

**Comment by etiennedi** [comment]
Please also note bullet 3 of https://github.com/semi-technologies/semi-website/issues/485#issuecomment-604365152 If the website requires a `.html` there should be redirect, as the problem itself should not have a `.html` ending.

**Comment by bobvanluijt** [comment]
> I wouldn't want to track it in a `yaml` file as it's quite complex to read in a yaml file at compile time in Go. However, I can generate a `yaml` file from the go code for the website if you like :)

Perfect, I'm agnostic about that one, it would just help us easily integrate it in the website (and test it of course).

@etiennedi could you already generate one with 1 or 2 error messages so that I can already integrate it in the website?

> Please also note bullet 3 of [#485 (comment)](https://github.com/semi-technologies/semi-website/issues/485#issuecomment-604365152) If the website requires a `.html` there should be redirect, as the problem itself should not have a `.html` ending.

Gotcha, this will be fine.
